### PR TITLE
8278330: dump stack trace if the jvmti test nsk/jvmti/GetThreadState/thrstat002 is failed with wrong thread state

### PR DESCRIPTION
**Pull request by sspitsyn**
This is a fix for sub-task of:
 8271065: vmTestbase/nsk/jvmti/GetThreadState/thrstat002 still fails with "Wrong thread "thr1" (...) state after SuspendThread"

The vmTestbase/nsk/jvmti/GetThreadState/thrstat002 test is intermittently failing with the wrong thread state reported.
It is most likely because of the unexpected JFR (or NMT) activity.
In order to confirm this guess we need to dump the stack trace in a case of this failure.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278330](https://bugs.openjdk.java.net/browse/JDK-8278330): dump stack trace if the jvmti test nsk/jvmti/GetThreadState/thrstat002 is failed with wrong thread state

### Reviewers
 * [Alex Menkov](https://openjdk.java.net/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Leonid Mesnik](https://openjdk.java.net/census#lmesnik) (@lmesnik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6735/head:pull/6735` \
`$ git checkout pull/6735`

Update a local copy of the PR: \
`$ git checkout pull/6735` \
`$ git pull https://git.openjdk.java.net/jdk pull/6735/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6735`

View PR using the GUI difftool: \
`$ git pr show -t 6735`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6735.diff">https://git.openjdk.java.net/jdk/pull/6735.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp [hunk 1]
```diff
@@ -48,6 +48,104 @@ static jint state[] = {
...
+static void
+print_method(jvmtiEnv *jvmti, JNIEnv* jni, jmethodID method, jint depth) {
+  char*  cname = NULL;
+  char*  mname = NULL;
+  char*  msign = NULL;
+  jvmtiError err;
+
+  cname = get_method_class_name(jvmti, jni, method);
+
+  err = jvmti->GetMethodName(method, &mname, &msign, NULL);
```
Comment (alexmenkov) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:111]: Need to deallocate returned mname/msig

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp [hunk 2]
```diff
@@ -48,6 +48,104 @@ static jint state[] = {
...
+get_method_class_name(jvmtiEnv *jvmti, JNIEnv* jni, jmethodID method) {
+  jclass klass = NULL;
+  char*  cname = NULL;
+  char*  result = NULL;
+  jvmtiError err;
+
+  err = jvmti->GetMethodDeclaringClass(method, &klass);
+  check_jvmti_status(jni, err, "get_method_class_name: error in JVMTI GetMethodDeclaringClass");
+
+  err = jvmti->GetClassSignature(klass, &cname, NULL);
```
Comment (alexmenkov) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:89]: Need to deallocate returned strings

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp [hunk 3]
```diff
@@ -48,6 +48,104 @@ static jint state[] = {
...
+}
+
+static void
+print_method(jvmtiEnv *jvmti, JNIEnv* jni, jmethodID method, jint depth) {
+  char*  cname = NULL;
+  char*  mname = NULL;
+  char*  msign = NULL;
+  jvmtiError err;
+
+  cname = get_method_class_name(jvmti, jni, method);
```
Comment (alexmenkov) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:109]: Need to deallocate returned string

**Review by alexmenkov — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp [hunk 4]
```diff
@@ -48,6 +48,104 @@ static jint state[] = {
     JVMTI_THREAD_STATE_IN_OBJECT_WAIT
 };
 
+#define DBG 1
```
Comment (lmesnik) [inline test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:51]: I don't think that DBG really needs to be disabled, just print info anytime when test fails

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp (hunk 4 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:51]: I used these ifdef's to easily identify the debugging/utility fragments in a case we want to isolate it in testing library. It does not harm anyway.

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp (hunk 3 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:109]: Thanks, fixed.

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp (hunk 2 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:88]: Thanks, fixed.

File: test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp (hunk 1 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jvmti/GetThreadState/thrstat002/thrstat002.cpp:111]: Thanks, fixed.

**Review by lmesnik — APPROVED** [review]
Looks good.

**Review by alexmenkov — APPROVED** [review]

**Pull request closed without merging**